docs/source/reference/benchmark/callback.rst

```diff
@@ -0,0 +1,130 @@
+
+.. code-block:: yaml
+
+    setup:
+        # Activate conda or virtualenv if you use one
+        # Then, install SkyCallback
+        pip install "git+https://github.com/skypilot-org/skypilot.git#egg=sky-callback&subdirectory=sky/callbacks/"
+
+
+Using SkyCallback generic APIs
+--------------------------------------------
+
+The SkyCallback generic APIs are for **PyTorch, TensorFlow, and JAX** programs where training loops are exposed to the users.
+Below we provide the instructions for using the APIs.
+
+First, import the SkyCallback package and initialize it using ``init``.
+You can optionally specify in ``total_steps`` the total number of steps that the task will iterate through.
+
+.. code-block:: python
+
+    import sky_callback
+    ...
+    sky_callback.init(
+        total_steps=num_epochs * len(train_dataloader), # Optional
+    )
+
+Next, mark the beginning and end of each step using one of the three equivalent methods.
+
+.. code-block:: python
+
+    # Method 1: wrap your iterable (e.g., dataloader) with `step_iterator`.
+    from sky_callback import step_iterator
+    for batch in step_iterator(train_dataloader):
+        ...
+
+    # Method 2: wrap your loop body with the `step` context manager.
+    for batch in train_dataloader:
+        with sky_callback.step():
+            ...
+
+    # Method 3: call `step_begin` and `step_end` directly.
+    for batch in train_dataloader:
+        sky_callback.step_begin()
+        ...
+        sky_callback.step_end()
+
+That's it.
+Now you can launch your task and get a complete benchmark report using SkyPilot Benchmark CLI.
+`Here <https://github.com/skypilot-org/skypilot/blob/master/examples/benchmark/timm/callback.patch>`_ we provide an example of applying SkyCallback to Pytorch ImageNet training code.
+
+.. note::
+    In distributed training, ``global_rank`` should be additionally passed to ``sky_callback.init`` as follows:
+
+    .. code-block:: python
+
+        # PyTorch DDP users
+        global_rank = torch.distributed.get_rank()
+
+        # Horovod users
+        global_rank = hvd.rank()
+
+        sky_callback.init(
+            global_rank=global_rank,
+            total_steps=num_epochs * len(train_dataloader), # Optional
+        )
+
+Integrations with ML frameworks
+----------------------------------------------------------
+
+Using SkyCallback is even easier for **Keras, PytorchLightning, and HuggingFace Transformers** programs where trainer APIs are used.
+SkyCallback natively supports these frameworks with simple interface.
+
+* Keras example
+
+.. code-block:: python
+
+    from sky_callback import SkyKerasCallback
+
+    # Add the callback to your Keras model.
+    model.fit(..., callbacks=[SkyKerasCallback()])
+
+`Here <https://github.com/skypilot-org/skypilot/blob/master/examples/benchmark/keras_asr/callback.patch>`_ you can find an example of applying SkyCallback to Keras ASR model training.
```

Comment:
Git patches, while convenient to use, are a little hard to read. Is there some other way we can show this code? Maybe a link to a fork with the patch applied? What do you think?

Reply:
I thought showing the full training code is inefficient as we need only two lines of code changes. I believe most users will understand the APIs without seeing the examples, and the examples are only to confirm that no other changes are required.